File: src/types/project.ts
/**
 * Контракт project.json — schemaVersion: 1
 * Неразрушающий проект записи с таймлайном и настройками рендера.
 */

export const PROJECT_SCHEMA_VERSION = 1 as const;

// --- Примитивы ---

/** Прямоугольник в нормализованных координатах [0.0–1.0]. */
export interface NormalizedRect {
  x: number;
  y: number;
  width: number;
  height: number;
}

// --- Таймлайн ---

export interface PanKeyframe {
  ts: number;
  offsetX: number;
  offsetY: number;
}

export interface TargetPoint {
  ts: number;
  rect: NormalizedRect;
}

export interface CameraSpring {
  mass: number;
  stiffness: number;
  damping: number;
}

export type ZoomMode = "fixed" | "follow-cursor";
export type ZoomTrigger = "auto-click" | "auto-scroll" | "manual";

export interface ZoomSegment {
  id: string;
  /** Начало сегмента (мс от начала записи). */
  startTs: number;
  /** Конец сегмента (мс). */
  endTs: number;
  /** Целевая область просмотра (нормализованные координаты). */
  initialRect?: NormalizedRect;
  targetRect?: NormalizedRect;
  targetPoints?: TargetPoint[];
  spring?: CameraSpring;
  panTrajectory?: PanKeyframe[];
  mode?: ZoomMode;
  trigger?: ZoomTrigger;
  /** true — создан алгоритмом авто-зума; false — пользователем вручную. */
  isAuto: boolean;
}

export interface Timeline {
  zoomSegments: ZoomSegment[];
}

// --- Настройки ---

export interface CursorSettings {
  /** Относительный размер, 1.0 = нормальный. */
  size: number;
  color: string;
  /** [0.0, 1.0] — сила сглаживания траектории. */
  smoothingFactor: number;
}

export type Background =
  | { type: "solid"; color: string }
  | { type: "gradient"; from: string; to: string; direction: string };

export interface ExportSettings {
  width: number;
  height: number;
  fps: number;
  codec: "h264" | "h265" | "vp9";
}

export interface ProjectSettings {
  cursor: CursorSettings;
  background: Background;
  export: ExportSettings;
}

// --- Корневой объект ---

export interface Project {
  schemaVersion: typeof PROJECT_SCHEMA_VERSION;
  id: string;
  name: string;
  /** Unix timestamp (мс) создания проекта. */
  createdAt: number;
  /** Путь к сырому видеофайлу относительно папки проекта. */
  videoPath: string;
  /** Путь к прокси-видео для монтажа (опционально). */
  proxyVideoPath?: string;
  /** Путь к events.json относительно папки проекта. */
  eventsPath: string;
  /** Длительность записи (мс). */
  durationMs: number;
  videoWidth: number;
  videoHeight: number;
  timeline: Timeline;
  settings: ProjectSettings;
}

// --- Фабрики / дефолты ---

export function defaultCursorSettings(): CursorSettings {
  return { size: 1.0, color: "#FFFFFF", smoothingFactor: 0.8 };
}

export function defaultBackground(): Background {
  return { type: "solid", color: "#1a1a2e" };
}

export function defaultExportSettings(): ExportSettings {
  return { width: 1920, height: 1080, fps: 30, codec: "h264" };
}

export function defaultProjectSettings(): ProjectSettings {
  return {
    cursor: defaultCursorSettings(),
    background: defaultBackground(),
    export: defaultExportSettings(),
  };
}

export function createProject(
  id: string,
  name: string,
  videoPath: string,
  eventsPath: string,
  videoWidth: number,
  videoHeight: number,
  durationMs: number
): Project {
  return {
    schemaVersion: PROJECT_SCHEMA_VERSION,
    id,
    name,
    createdAt: Date.now(),
    videoPath,
    eventsPath,
    durationMs,
    videoWidth,
    videoHeight,
    timeline: { zoomSegments: [] },
    settings: defaultProjectSettings(),
  };
}

File: src-tauri/src/models/project.rs
//! Схема проекта (project.json).
//! schemaVersion: 1

use serde::{Deserialize, Serialize};

pub const SCHEMA_VERSION: u32 = 1;

/// Прямоугольная область в нормализованных координатах (0.0–1.0).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NormalizedRect {
    pub x: f64,
    pub y: f64,
    pub width: f64,
    pub height: f64,
}

fn default_normalized_rect() -> NormalizedRect {
    NormalizedRect {
        x: 0.0,
        y: 0.0,
        width: 1.0,
        height: 1.0,
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PanKeyframe {
    pub ts: u64,
    pub offset_x: f64,
    pub offset_y: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TargetPoint {
    pub ts: u64,
    pub rect: NormalizedRect,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CameraSpring {
    pub mass: f64,
    pub stiffness: f64,
    pub damping: f64,
}

impl Default for CameraSpring {
    fn default() -> Self {
        Self {
            mass: 1.0,
            stiffness: 170.0,
            damping: 26.0,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub enum ZoomMode {
    Fixed,
    FollowCursor,
}

impl Default for ZoomMode {
    fn default() -> Self {
        Self::Fixed
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub enum ZoomTrigger {
    AutoClick,
    AutoScroll,
    Manual,
}

impl Default for ZoomTrigger {
    fn default() -> Self {
        Self::Manual
    }
}

/// Один зум-сегмент на таймлайне.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ZoomSegment {
    pub id: String,
    /// Начало сегмента (мс от начала записи).
    pub start_ts: u64,
    /// Конец сегмента (мс).
    pub end_ts: u64,
    /// Целевая область просмотра (нормализованные координаты).
    #[serde(default = "default_normalized_rect", alias = "targetRect")]
    pub initial_rect: NormalizedRect,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub target_points: Vec<TargetPoint>,
    #[serde(default)]
    pub spring: CameraSpring,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    #[serde(alias = "panTrajectory")]
    pub pan_trajectory: Vec<PanKeyframe>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[serde(alias = "easing")]
    pub legacy_easing: Option<String>,
    #[serde(default)]
    pub mode: ZoomMode,
    #[serde(default)]
    pub trigger: ZoomTrigger,
    /// true — создан алгоритмом, false — пользователем вручную.
    #[serde(default)]
    pub is_auto: bool,
}

/// Таймлайн проекта.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Timeline {
    pub zoom_segments: Vec<ZoomSegment>,
}

/// Настройки курсора.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CursorSettings {
    /// Относительный размер курсора (1.0 = нормальный).
    pub size: f64,
    pub color: String,
    /// 0.0 = нет сглаживания, 1.0 = максимальное.
    pub smoothing_factor: f64,
}

impl Default for CursorSettings {
    fn default() -> Self {
        CursorSettings {
            size: 1.0,
            color: "#FFFFFF".to_string(),
            smoothing_factor: 0.8,
        }
    }
}

/// Тип фона.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum Background {
    Solid {
        color: String,
    },
    Gradient {
        from: String,
        to: String,
        direction: String,
    },
}

impl Default for Background {
    fn default() -> Self {
        Background::Solid {
            color: "#1a1a2e".to_string(),
        }
    }
}

/// Настройки экспорта.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExportSettings {
    pub width: u32,
    pub height: u32,
    pub fps: u32,
    pub codec: String,
}

impl Default for ExportSettings {
    fn default() -> Self {
        ExportSettings {
            width: 1920,
            height: 1080,
            fps: 30,
            codec: "h264".to_string(),
        }
    }
}

/// Настройки проекта.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ProjectSettings {
    pub cursor: CursorSettings,
    pub background: Background,
    pub export: ExportSettings,
}

/// Корневой объект project.json.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Project {
    pub schema_version: u32,
    pub id: String,
    pub name: String,
    /// Unix timestamp (мс) создания проекта.
    pub created_at: u64,
    /// Путь к сырому видеофайлу относительно папки проекта.
    pub video_path: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub proxy_video_path: Option<String>,
    /// Путь к файлу событий относительно папки проекта.
    pub events_path: String,
    /// Длительность записи (мс).
    pub duration_ms: u64,
    /// Разрешение захваченного видео.
    pub video_width: u32,
    pub video_height: u32,
    pub timeline: Timeline,
    pub settings: ProjectSettings,
}
